datafusion/sql/src/unparser/utils.rs

```diff
@@ -78,22 +103,33 @@ pub(crate) fn find_agg_node_within_select<'a>(
 ///
 /// For example, if expr contains the column expr "COUNT(*)" it will be transformed
 /// into an actual aggregate expression COUNT(*) as identified in the aggregate node.
-pub(crate) fn unproject_agg_exprs(expr: &Expr, agg: &Aggregate) -> Result<Expr> {
+pub(crate) fn unproject_agg_exprs(
+    expr: &Expr,
+    agg: &Aggregate,
+    windows: Option<&[&Window]>,
+) -> Result<Expr> {
     expr.clone()
         .transform(|sub_expr| {
             if let Expr::Column(c) = sub_expr {
-                // find the column in the agg schema
-                if let Ok(n) = agg.schema.index_of_column(&c) {
-                    let unprojected_expr = agg
-                        .group_expr
-                        .iter()
-                        .chain(agg.aggr_expr.iter())
-                        .nth(n)
-                        .unwrap();
+                if let Some(unprojected_expr) = find_agg_expr(agg, &c)? {
                     Ok(Transformed::yes(unprojected_expr.clone()))
+                } else if let Some(mut unprojected_expr) =
+                    windows.and_then(|w| find_window_expr(w, &c.name).cloned())
+                {
+                    if let Expr::WindowFunction(func) = &mut unprojected_expr {
+                        // Window function can contain an aggregation column, e.g., 'avg(sum(ss_sales_price)) over ...' that needs to be unprojected
+                        for arg in &mut func.args {
+                            if let Expr::Column(c) = arg {
+                                if let Some(expr) = find_agg_expr(agg, c)? {
+                                    *arg = expr.clone();
+                                }
+                            }
+                        }
```

Comment:
We could consider the following style. I prefer to use `iter_mut` instead.
```suggestion
                        func.args.iter_mut().try_for_each(|arg| {
                            if let Expr::Column(c) = arg {
                                if let Some(expr) = find_agg_expr(agg, c)? {
                                    *arg = expr.clone();
                                }
                            }
                            Ok::<(), DataFusionError>(())
                        })?;
                    }
```